### PR TITLE
fix for nvme device path

### PR DESCRIPTION
**Pull request by youngtack**
nvme device path doesn't have "block" as like "nvme/nvme0/nvme0n1"
or "nvme/nvme0/nvme0n1/nvme0n1p1". the last directory such as
"nvme0n1p1" should be removed if nvme drive is partitioned.

### CODE REVIEW

File: env/io_posix.cc [hunk 1]
```diff
@@ -95,8 +95,9 @@ size_t GetLogicalBufferSize(int __attribute__((__unused__)) fd) {
   if (parent_begin == std::string::npos) {
     return kDefaultPageSize;
   }
-  if (device_dir.substr(parent_begin + 1, parent_end - parent_begin - 1) !=
-      "block") {
+  std::string parent = device_dir.substr(parent_begin + 1, parent_end - parent_begin - 1);
+  std::string child = device_dir.substr(parent_end + 1, device_dir.length() - 1);
+  if (parent != "block" && (child.compare(0, 4, "nvme") || child.find('p') != std::string::npos)) {
```
Comment (riversand963) [inline env/io_posix.cc]: If `parent != "block"` AND `child.compare(0, 4, "nvme") == 0` (starts with "nvme"), then we have to evaluate `child.find('p')`. If child has 'p', then we are going to execute inside the `if`. In this case, does it mean it's not an nvme even if child starts with "nvme"?

**Comment by riversand963** [comment]
Can you also provide more explanation in the PR's description?

File: env/io_posix.cc [hunk 2]
```diff
@@ -95,8 +95,9 @@ size_t GetLogicalBufferSize(int __attribute__((__unused__)) fd) {
   if (parent_begin == std::string::npos) {
     return kDefaultPageSize;
   }
-  if (device_dir.substr(parent_begin + 1, parent_end - parent_begin - 1) !=
-      "block") {
+  std::string parent = device_dir.substr(parent_begin + 1, parent_end - parent_begin - 1);
+  std::string child = device_dir.substr(parent_end + 1, device_dir.length() - 1);
```
[the remark this answers is not in the document]
Comment (youngtack) [reply env/io_posix.cc]: `npos` looks clearer than `length() - 1`. But I can't figure out how to amend this pull request. Could you advise how-to or amend this?

File: env/io_posix.cc (hunk 1 above)
Comment (youngtack) [reply env/io_posix.cc]: just to remove the last child, `"nvme0n1p1"` out of `"nvme/nvme0/nvme0n1/nvme0n1p1"` if nvme drive is partitioned.

File: env/io_posix.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (riversand963) [reply env/io_posix.cc]: You can just checkout this current branch `patch-1`, make changes, `git commit -a`, then `git push origin patch-1:patch-1`. This should work in most cases.

File: env/io_posix.cc (hunk 1 above)
Comment (riversand963) [reply env/io_posix.cc]: Got it, thanks

File: env/io_posix.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (youngtack) [reply env/io_posix.cc]: Thanks a lot for your kind advice. Great to see the forked branch is applied to mainline through git CLI.

File: env/io_posix.cc (hunk 1 above)
Comment (youngtack) [reply env/io_posix.cc]: Good question. This below would explain partitioned nvme drive naming convention as I just crawled into linux kernel source code.
```
linux-stable/block/partition-generic.c
struct hd_struct *add_partition(struct gendisk *disk, int partno,
                                sector_t start, sector_t len, int flags,
                                struct partition_meta_info *info)
{
...
        dname = dev_name(ddev);
        if (isdigit(dname[strlen(dname) - 1]))
                dev_set_name(pdev, "%sp%d", dname, partno);
        else
                dev_set_name(pdev, "%s%d", dname, partno);
...
}
```

**Comment by youngtack** [comment]
Hi @riversand963,

So long after this PR is revised according to your comment. Could you tell me when this would be merged?

Thanks,
@youngtack

**Comment by riversand963** [comment]
@youngtack , sure. Would you mind updating the description of the PR, so that other people can more easily understand? Thanks.

btw. I just pushed a commit to adjust the format.

**Comment by youngtack** [comment]
@riversand963: Thanks to adjust the format. I revised PR description and please have a look into it.

**Comment by riversand963** [comment]
Thanks @youngtack .
Actually I just dug more into this issue, and have some concern.
According to https://www.intel.com/content/dam/support/us/en/documents/ssdc/data-center-ssds/Intel_Linux_NVMe_Guide_330602-002.pdf, it seems that `/sys/block/nvme*n` is possible. There are also other examples in which I can see `/sys/block/nvme*n`, e.g. https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1778765.
In https://elixir.bootlin.com/linux/v4.20.5/source/drivers/nvme/host/core.c#L1320, it seems to suggest that the device is treated as a block device and should appear in `/sys/block`. I am not very familiar with most update to date nvme linux driver, can you provide more background on this? Does your PR target specific kernel distributions and versions? Thanks!

**Comment by youngtack** [comment]
@riversand963: Thanks to investigate linux documents for this PR. Trimming the device path is done only when the parent is not "block" and the child has 'p'. Trimming is never done if the parent is "block" or the child doesn't have 'p'. So the cases you found would have no problem to reach logical_block_size file with this patch.  I verified this patch on kernel-ml-4.{19,20}.x of CentOS known to be based on linux-stable repo. Thanks.

**Pull request closed without merging**